File: Binary Search/2187.Minimum Time to Complete Trips/Solution.java
class Solution {
    public long minimumTime(int[] time, int totalTrips) {
        long minTime = Integer.MAX_VALUE;
        for (int t : time) {
            minTime = Math.min(minTime, t);
        }
        long left = 0, right = minTime * totalTrips;
        while (left < right) {
            long mid = (left + right) >>> 1;
            if (check(time, mid, totalTrips)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }

    public boolean check(int[] time, long mid, int totalTrips) {
        long cnt = 0;
        for (int t : time) {
            cnt += mid / t;
        }
        return cnt >= totalTrips;
    }
}

File: Binary Search/2187.Minimum Time to Complete Trips/Solution.py
class Solution:
    def minimumTime(self, time: List[int], totalTrips: int) -> int:
        def check(k):
            nonlocal totalTrips, time
            cnt = 0
            for t in time:
                cnt += k // t
            return cnt >= totalTrips

        min_time = min(time)
        left, right = 0, min_time * totalTrips
        while left < right:
            mid = (left + right) >> 1
            if check(mid):
                right = mid
            else:
                left = mid + 1
        return left